File: src/ts/app/BlogPost.ts
import BlogUrl from "./BlogUrl";

class BlogPost {
    private readonly _id: string;
    private readonly _commentsEnabled: boolean;
    private readonly _title: string;
    private readonly _excerpt: string;
    private readonly _content: string;
    private readonly _authorId: string;
    private readonly _published: Date;
    private readonly _updated?: Date;
    private readonly _url: BlogUrl;
    private readonly _trimmed: boolean;
    private readonly _identifier: string;
    private readonly _humanizedTimestamp: string;
    private readonly _formattedPublishDate: string;
    private readonly _formattedUpdateDate: string;
    private readonly _tags: string[];

    constructor(json: any) {
        this._id = json.id;
        this._commentsEnabled = json.commentsEnabled;
        this._title = json.title;
        this._excerpt = json.excerpt;
        this._content = json.content;
        this._authorId = json.author;
        this._published = new Date(json.published * 1000);
        this._updated = (json.updated && new Date(json.updated * 1000)) || null;
        this._url = new BlogUrl(json.url);
        this._trimmed = json.trimmed;
        this._identifier = json.identifier;
        this._humanizedTimestamp = json.humanizedTimestamp;
        this._formattedPublishDate = json.formattedPublishDate;
        this._formattedUpdateDate = json.formattedUpdateDate;
        this._tags = json.tags;
    }

    get id(): string {
        return this._id;
    }

    get commentsEnabled(): boolean {
        return this._commentsEnabled;
    }

    get title(): string {
        return this._title;
    }

    get excerpt(): string {
        return this._excerpt;
    }

    get content(): string {
        return this._content;
    }

    get authorId(): string {
        return this._authorId;
    }

    get published(): Date {
        return this._published;
    }

    get updated(): Date {
        return this._updated;
    }

    get url(): BlogUrl {
        return this._url;
    }
    
    get tags(): string[] {
        return this._tags;
    }

    get trimmed(): boolean {
        return this._trimmed;
    }

    get identifier(): string {
        return this._identifier;
    }

    get humanizedTimestamp(): string {
        return this._humanizedTimestamp;
    }

    get formattedPublishDate(): string {
        return this._formattedPublishDate;
    }

    get formattedUpdateDate(): string {
        return this._formattedUpdateDate;
    }
}

export default BlogPost;

File: OliverBooth.Api/Data/BlogPost.cs
using System.Text.Json.Serialization;
using Humanizer;
using OliverBooth.Common.Data.Blog;
using OliverBooth.Common.Services;

namespace OliverBooth.Api.Data;

internal sealed class BlogPost
{
    [JsonPropertyName("author"), JsonInclude, JsonPropertyOrder(3)]
    public Guid Author { get; private set; }

    [JsonPropertyName("commentsEnabled"), JsonInclude, JsonPropertyOrder(1)]
    public bool CommentsEnabled { get; private set; }

    [JsonPropertyName("content"), JsonInclude, JsonPropertyOrder(11)]
    public string? Content { get; private set; }

    [JsonPropertyName("excerpt"), JsonInclude, JsonPropertyOrder(10)]
    public string Excerpt { get; private set; } = string.Empty;

    [JsonPropertyName("formattedPublishDate"), JsonInclude, JsonPropertyOrder(7)]
    public string FormattedPublishDate { get; private set; } = string.Empty;

    [JsonPropertyName("formattedUpdateDate"), JsonInclude, JsonPropertyOrder(8)]
    public string? FormattedUpdateDate { get; private set; }

    [JsonPropertyName("humanizedTimestamp"), JsonInclude, JsonPropertyOrder(9)]
    public string HumanizedTimestamp { get; private set; } = string.Empty;

    [JsonPropertyName("id"), JsonInclude, JsonPropertyOrder(0)]
    public Guid Id { get; private set; }

    [JsonPropertyName("identifier"), JsonInclude, JsonPropertyOrder(2)]
    public string Identifier { get; private set; } = string.Empty;

    [JsonPropertyName("trimmed"), JsonInclude, JsonPropertyOrder(12)]
    public bool IsTrimmed { get; private set; }

    [JsonPropertyName("published"), JsonInclude, JsonPropertyOrder(5)]
    public long Published { get; private set; }

    [JsonPropertyName("tags"), JsonInclude, JsonPropertyOrder(13)]
    public IEnumerable<string> Tags { get; private set; } = ArraySegment<string>.Empty;

    [JsonPropertyName("title"), JsonInclude, JsonPropertyOrder(4)]
    public string Title { get; private set; } = string.Empty;

    [JsonPropertyName("updated"), JsonInclude, JsonPropertyOrder(6)]
    public long? Updated { get; private set; }

    [JsonPropertyName("url"), JsonInclude, JsonPropertyOrder(14)]
    public object Url { get; private set; } = null!;

    public static BlogPost FromBlogPost(IBlogPost post, IBlogPostService blogPostService,
        bool includeContent = false)
    {
        return new()
        {
            Id = post.Id,
            CommentsEnabled = post.EnableComments,
            Identifier = post.GetDisqusIdentifier(),
            Author = post.Author.Id,
            Title = post.Title,
            Published = post.Published.ToUnixTimeSeconds(),
            Updated = post.Updated?.ToUnixTimeSeconds(),
            FormattedPublishDate = post.Published.ToString("dddd, d MMMM yyyy HH:mm"),
            FormattedUpdateDate = post.Updated?.ToString("dddd, d MMMM yyyy HH:mm"),
            HumanizedTimestamp = post.Updated?.Humanize() ?? post.Published.Humanize(),
            Excerpt = blogPostService.RenderExcerpt(post, out bool trimmed),
            Content = includeContent ? blogPostService.RenderPost(post) : null,
            IsTrimmed = trimmed,
            Tags = post.Tags.Select(t => t.Replace(' ', '-')),
            Url = new
            {
                year = post.Published.ToString("yyyy"),
                month = post.Published.ToString("MM"),
                day = post.Published.ToString("dd"),
                slug = post.Slug
            }
        };
    }
}
